gtsfm/utils/metrics.py

```diff
@@ -60,16 +66,147 @@ def count_correct_correspondences(
+
+    Returns:
+        is_inlier: (N, ) mask of inlier correspondences.
+        reproj_err: maximum error between forward-projected ground truth landmark and corresponding keypoints
+
+    Raises:
+        ValueError if the number of keypoints do not match.
+    """
+    if len(keypoints_i1) != len(keypoints_i2):
+        raise ValueError("Keypoints must have same counts")
+    n_corrs = len(keypoints_i1)
+    is_inlier = np.zeros(n_corrs, dtype=bool)
+
+    # Perform ray tracing to compute keypoint intersections.
+    keypoint_ind_i1, intersections_i1 = compute_keypoint_intersections(keypoints_i1, gt_camera_i1, gt_scene_mesh)
+    keypoint_ind_i2, intersections_i2 = compute_keypoint_intersections(keypoints_i2, gt_camera_i2, gt_scene_mesh)
+    keypoint_ind, i1_idx, i2_idx = np.intersect1d(keypoint_ind_i1, keypoint_ind_i2, return_indices=True)
+
+    # Forward project intersections into other image to compute error.
+    reproj_err = np.array([np.nan] * len(keypoints_i1))
+    for i in range(len(keypoint_ind)):
+        uv_i1 = keypoints_i1.coordinates[keypoint_ind[i]]
+        uv_i2 = keypoints_i2.coordinates[keypoint_ind[i]]
+        uv_i2i1, success_flag_i1 = gt_camera_i1.projectSafe(intersections_i2[i2_idx[i]])
+        uv_i1i2, success_flag_i2 = gt_camera_i2.projectSafe(intersections_i1[i1_idx[i]])
+        if success_flag_i1 and success_flag_i2:
+            err_i2i1 = np.linalg.norm(uv_i1.flatten() - uv_i2i1.flatten())
+            err_i1i2 = np.linalg.norm(uv_i2.flatten() - uv_i1i2.flatten())
+            is_inlier[keypoint_ind[i]] = max(err_i2i1, err_i1i2) < dist_threshold
+            reproj_err[keypoint_ind[i]] = max(err_i2i1, err_i1i2)
+        else:
+            is_inlier[keypoint_ind[i]] = False
+            reproj_err[keypoint_ind[i]] = np.nan
+
+    return is_inlier, reproj_err
+
+
+def compute_keypoint_intersections(
+    keypoints: Keypoints, gt_camera: PinholeCameraCal3Bundler, gt_scene_mesh: trimesh.Trimesh
+) -> Tuple[np.ndarray, np.ndarray]:
+    """Computes intersections between ground truth surface mesh and rays originating from image keypoints.
+
+    Args:
+        keypoints: N keypoints computed in image.
+        gt_camera: ground truth camera.
+        gt_scene_mesh: ground truth triangular surface mesh.
+
+    Returns:
+        keypoint_ind: (M,) array of keypoint indices whose corresponding ray intersected the ground truth mesh.
+        intersections_locations: (M, 3), array of ray intersection locations.
+    """
+    num_kpts = len(keypoints)
+    src = np.repeat(gt_camera.pose().translation().reshape((-1, 3)), num_kpts, axis=0)  # At_i1A
+    drc = np.asarray([gt_camera.backproject(keypoints.coordinates[i], depth=1.0) - src[i, :] for i in range(num_kpts)])
+    start_time = timeit.default_timer()
+    intersections, keypoint_ind, _ = gt_scene_mesh.ray.intersects_location(src, drc, multiple_hits=False)
+    logger.info(f"Case {num_kpts} rays in {timeit.default_timer() - start_time} seconds.")
```

Comment:
Still a bit hesitant about logging this per image pair, since we could be printing out 100,000s of lines with this call here on large datasets.

Reply:
we should start using logging levels instead of unnecessary arguments: https://docs.python.org/3/library/logging.html#logging.Logger.log